card/lib/card/auth.rb

```diff
@@ -135,12 +139,12 @@ def as given_user
         @@current_id = @@as_id if @@current_id.nil?
 
         return unless block_given?
-        value = yield
-
-        @@as_id   = tmp_id
-        @@as_card = tmp_card
-
-        value
+        yield
+      ensure
+        if block_given?
+          @@as_id   = tmp_id
+          @@as_card = tmp_card
+        end
```

Comment:
seems safer to me to `ensure` that we return after the block to the previous settings